### PR TITLE
Improve documentation and add examples

### LINKED ISSUE #2: Document public items
It would be great if these public items are documented, especially for the tuple `finalized()` returning.

https://github.com/rust-lang/rustc-stable-hash/blob/cb8e141b08fb839606a5f79f9b56087cd54b764d/src/stable_hasher.rs#L33-L50

The implementation of `Hash::finish` should also be documented, with the reason why it can't be used.

https://github.com/rust-lang/rustc-stable-hash/blob/cb8e141b08fb839606a5f79f9b56087cd54b764d/src/stable_hasher.rs#L52-L55

Would be great if we can provide a reference implementation of `StableHasherResult` trait, or link to the implementation of `Hash64`/`Hash128` in rust-lang/rust.

https://github.com/rust-lang/rustc-stable-hash/blob/cb8e141b08fb839606a5f79f9b56087cd54b764d/src/stable_hasher.rs#L11-L14

### PR DESCRIPTION
**Pull request by Urgau**
This PR improves the documentation of the public methods of `StableHasher` and adds some usage examples.

cc @weihanglo
r? @WaffleLapkin
Fixes #2

### CODE REVIEW

File: src/stable_hasher.rs [hunk 1]
```diff
@@ -8,48 +8,113 @@ use std::hash::Hasher;
...
+    #[must_use]
     pub fn finish<W: StableHasherResult>(self) -> W {
         W::finish(self)
     }
 
+    /// Returns the hash value for the values written so far.
+    ///
+    /// This method returns a `(u64, u64)` pair for greater precision.
+    ///
+    /// See [`StableHasher::finish`] for an idomatic use of this function.
```
Comment (WaffleLapkin) [inline src/stable_hasher.rs]: this point doesn't make sense to me. if you use `finish` you don't just this function...

File: src/stable_hasher.rs [hunk 2]
```diff
@@ -8,48 +8,113 @@ use std::hash::Hasher;
...
 impl Hasher for StableHasher {
+    /// <div class="warning">
+    ///
+    /// Do not use this function, it will unconditionnaly panic.
+    ///
+    /// Use instead [`StableHasher::finish`], or [`StableHasher::finalize`]
+    /// which returns a `(u64, u64)` pair for greater precision.
+    ///
+    /// </div>
     fn finish(&self) -> u64 {
```
Comment (WaffleLapkin) [inline src/stable_hasher.rs:108]: Does `deprecated` work on implementations? If so, you could add it here.

File: src/stable_hasher.rs [hunk 3]
```diff
@@ -8,48 +8,113 @@ use std::hash::Hasher;
...
+    }
+}
+
 impl Hasher for StableHasher {
+    /// <div class="warning">
+    ///
+    /// Do not use this function, it will unconditionnaly panic.
+    ///
+    /// Use instead [`StableHasher::finish`], or [`StableHasher::finalize`]
+    /// which returns a `(u64, u64)` pair for greater precision.
```
Comment (WaffleLapkin) [inline src/stable_hasher.rs]: hm, but what if we don't need that much precision? iirc there was `Hash64` structure in rustc...

Comment (Urgau) [reply src/stable_hasher.rs]: `Hash64` just use the "lower" part in it's [implementation](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_data_structures/hashes.rs.html#61-63) of `StableHasherResult` and ignores the upper part.

I think the goal with `StableHasher::finish` is to be explicit about the use (or non use) of part of result hash and since we can't return anything else than `u64` with the default `Hasher::finish`, we have `StableHasher::{finish,finalize}` methods to do that.

In other words making `Hasher::finish` only return the lower part without the upper part is unexpected, so instead we just try to make it explicit by "forcing" the usage of `StableHasher::{finish,finalize}`.

File: src/stable_hasher.rs (hunk 2 above)
Comment (Urgau) [reply src/stable_hasher.rs:108]: Unfortunately no,
```
error: this `#[deprecated]` annotation has no effect
   --> src/stable_hasher.rs:118:5
    |
118 |     #[deprecated = "use StableHasher::finalize instead"]
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove the unnecessary deprecation attribute
    |
    = note: `#[deny(useless_deprecated)]` on by default

error: could not compile `rustc-stable-hash` (lib) due to 1 previous error
```

File: src/stable_hasher.rs (hunk 1 above)
Comment (Urgau) [reply src/stable_hasher.rs]: It's used indirectly. Calling `StableHasher::finish` will call `StableHasherResult::finish` which should call `StableHasher::finalize`.

Comment (Urgau) [reply src/stable_hasher.rs]: I looked all 3 implementation of `StableHasherResult` in `rustc` and they all immediately call `StableHasher::finalize` in their `StableHasherResult::finish` implementation, we could therefore conceivably remove `StableHasher::finalize` and directly pass the `(u64, u64)` tuple in `StableHasherResult::finish`; which would make this less confusing and simpler to understand. Thoughs?

Comment (weihanglo) [reply src/stable_hasher.rs]: Should we also call out that `.0` is the lower part and `.1` is the upper part?

Comment (michaelwoerister) [reply src/stable_hasher.rs]: Could I still do `let hash: SomethingImplemetingStableHasherResult = hasher.finish();`?

Comment (Urgau) [reply src/stable_hasher.rs]: > Could I still do `let hash: SomethingImplemetingStableHasherResult = hasher.finish();`?

Yes. The only (other) thing that will change if we remove `StableHasher::finalize` is the signature of `StableHasherResult::finish`:
```diff
- fn finish(hasher: StableHasher) -> Self;
+ fn finish(hash: (u64, u64)) -> Self;
```

That would remove the need for implementors of `StableHasherResult` to manually call `StableHasher::finalize` and simplify the interface.

Comment (michaelwoerister) [reply src/stable_hasher.rs]: Is this what you mean?

```rust
pub trait StableHasherResult: Sized {
    fn from(hash: (u64, u64)) -> Self;
}
```

Comment (Urgau) [reply src/stable_hasher.rs]: > Is this what you mean?

Yes

Comment (michaelwoerister) [reply src/stable_hasher.rs]: Whoops, I didn't see your reply 🙂
That seems like a fine change to me.

File: src/stable_hasher.rs (hunk 2 above)
Comment (michaelwoerister) [reply src/stable_hasher.rs:108]: Afaik, the panic is from a time when there was no `StableHashResult` yet. It should be fine to just return one half of the hash or do something like [Fingerprint::combine](https://github.com/rust-lang/rust/blob/1aaab8b9f8dc488cadc4f083b3a11fb11b45cb77/compiler/rustc_data_structures/src/fingerprint.rs#L44-L51).

File: src/stable_hasher.rs (hunk 3 above)
Comment (weihanglo) [reply src/stable_hasher.rs]: https://github.com/rust-lang/rustc-stable-hash/blob/cb8e141b08fb839606a5f79f9b56087cd54b764d/src/sip128.rs#L3-L5

Given this is stated as “hot”, will we gain any benefit if we had a specialized 64bit version of this stable hasher?

Comment (michaelwoerister) [reply src/stable_hasher.rs]: As far as I know the two versions are basically the same except for how the final hash value is computed from the hasher's state at the end. So I don't think so.

File: src/stable_hasher.rs (hunk 1 above)
Comment (WaffleLapkin) [reply src/stable_hasher.rs]: > It's used indirectly. 

yes, but "idiomatic use" implied that you actually call the function directly.

the above change to the result trait to accept a `[u64; 2]` makes sense to me.

Comment (Urgau) [reply src/stable_hasher.rs]: > the above change to the result trait to accept a `[u64; 2]` makes sense to me.

I posted https://github.com/rust-lang/rustc-stable-hash/pull/4 to address this.

**Comment by Urgau** [comment]
#4 was merged, I have rebased this PR over `main`, CI is green and I think https://github.com/rust-lang/rustc-stable-hash/pull/3#discussion_r1647713568 can be handled as a follow-up (rather than holding this PR). I think this is ready be merged.

@WaffleLapkin @michaelwoerister any objections?

File: src/stable_hasher.rs [hunk 4]
```diff
@@ -8,43 +8,100 @@ use std::hash::Hasher;
...
+/// struct Hash128([u64; 2]);
+///
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hash: [u64; 2]) -> Hash128 {
+///         Hash128(hash)
+///     }
+/// }
+/// ```
+pub trait StableHasherResult: Sized {
+    /// Retriving the finalized state of the [`StableHasher`] and construct
```
Comment (michaelwoerister) [inline src/stable_hasher.rs]:
```suggestion
    /// Retrieving the finalized state of the [`StableHasher`] and construct
```

File: src/stable_hasher.rs [hunk 5]
```diff
@@ -8,43 +8,100 @@ use std::hash::Hasher;
...
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+///
+/// struct Hash128([u64; 2]);
+///
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hash: [u64; 2]) -> Hash128 {
+///         Hash128(hash)
+///     }
+/// }
+/// ```
```
Comment (michaelwoerister) [inline src/stable_hasher.rs:61]: The example could does something more interesting to make it clear why a transformation step might even be needed. E.g. the `Hash128` in the example could contain a `u128`.

**Comment by michaelwoerister** [comment]
@Urgau, I left a couple of remarks but nothing blocking. I'm good with merging the PR, looks great!

File: src/stable_hasher.rs (hunk 5 above)
Comment (Urgau) [reply src/stable_hasher.rs:61]: Indeed, that's a good observation, thks. I've updated the `StableHasherResult` example to be `Hash128(u128)`, as to make it more *interesting*.

**Review by WaffleLapkin — APPROVED** [review]
